### PR TITLE
[RNMobile] Android Only - Fix Gboard enter detection in EnterPressedWatcher

### LINKED ISSUE #3316: Enter detection fails in EnterPressedWatcher for Gboard keyboard
## Description

The mechanism to detect whether the `[ENTER]` key was pressed is not reliable for users with the Gboard keyboard. The Gboard keyboard exhibits some unusual behavior regarding the `TextWatcher` interface methods, which are breaking some assumptions in the logic we are using for key detection. This manifests as various bugs, and is likely not limited to the `[ENTER]` key (e.g. I believe `[BACKSPACE]` detection may be affected in a similar manner).

### Example bug

On a Pixel 3a (physical device) with the Gboard keyboard:

1. Create a new post
2. Add a paragraph block
3. Type a word (e.g. "Hello") without a space
  * Observe that the keyboard suggestions
  * Observe that the word is underlined
4. Press `[ENTER]`
5. Observe that the block does not split - instead, a newline is added to the paragraph block

#### Expected

The block should split and form a new paragraph block below.

## Deeper details

The `EnterPressedWatcher` class works by implementing the `TextWatcher` interface, overriding the following methods:

* `beforeTextChanged`
* `onTextChanged`
* `afterTextChanged`

and tracking state during a user's manipulation of the `EditText` input. These methods allow us to observe the mutations to the `CharSequence`, and to provide a means to reconcile what has changed, and respond in some way after evaluating the changes at each stage of the update. Typically, these changes are fairly straightforward, and these methods are invoked with predictable index boundaries for a given text input operation, but here is where Gboard exhibits some unusual behavior.

Here is a table describing some observations that surprised me:

Consider a scenario where we have the input with text `Hello worl`, and we then press `d` on the keyboard:

Comparison |`TextWatcher` Interface Method | `text: CharSequence` | `start: Int` | `count: Int` | `after: Int` | `before: Int`
-|-|-|-|-|-|-
Expected | `beforeTextChanged` | "Hello worl" | 10 | 0 | 1 | N/A
Actual | `beforeTextChanged`| "Hello worl" | 6 | 4 | 5 | N/A
Expected | `onTextChanged` | "Hello world" | 10 | 1 | N/A | 0
Actual | `onTextChanged`| "Hello world" | 6 | 5 | N/A | 4

The surprising part is that some characters from the original text are being replaced when nothing was selected (just a normal caret at the end of the input). It seems instead that Gboard is actually replacing the last 4 characters with those same 4 characters _plus_ the typed text. It is even possible to notice a small underline style applied to the characters that will be included in the replacement. Before this, I would have only expected `beforeTextChanged` to be called with a `count` > 0 if there was a selection prior to the new characters being typed (or if characters were being deleted).

Note that the result is the same (and that the sum is still 11 for `start + after` and `start + count` for `beforeTextChanged` and `onTextChanged` respectively).

Interestingly, after selecting one of the suggestions offered by Gboard, the `onEnter` behavior was works normally (and the underlined text goes away).

### PR DESCRIPTION
**Pull request by AmandaRiu**
**Gutenberg-mobile PR:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3590

Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/3316. This PR is an updated version of [this older draft PR](https://github.com/WordPress/gutenberg/pull/30368) by @mkevins and uses a new branch created by @hypest. I tested the fixes on the following physical devices:

Phone | OS
-- | --
Pixel 2 | Android 10
Pixel 4 | Android 11

Ran through all these writing flow tests first using the default Gboard. I then installed the Microsoft Swift Keyboard and ran through the tests again. Both keyboard's failed **TC002** in the **Splitting and merging** group of test cases. The test that failed was the one where you first split a paragraph block, then add text to the beginning of the second block and then select that new text and delete it. Tapping delete again will not merge the two blocks back into a single block. 

I also tested all the **known issues** with the two keyboards. Only one existing issue was fixed **for both keyboards**: 
- **\[Android\]** Cannot split blocks immediately after prepending to a "swiped" word [wordpress-mobile/gutenberg-mobile#2372](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2372)

However, **all known issues** were not an issue for the Swift keyboard. 

## Description

The approach used in this PR adds logic to detect when we've encountered the unusual behavior exhibited by Gboard which leads to a failure to detect onEnter keypresses. When such behavior is detected, it calculates new offsets to be used in string comparison and processing to conform to the original intention of this TextWatcher.

Note: This is probably not enough, since this only addresses when the caret is at the end of the word, but Gboard may exhibit the unusual behavior also when the caret is in the middle or start of the word. Some consideration should be made for those other conditions as well.

## How has this been tested

[The **Writing Flow Tests** in a comment below](https://github.com/WordPress/gutenberg/pull/32471#issuecomment-855083670) should be completed using a physical device with a and without a GBoard keyboard:

- [x] Physical device using the GBoard keyboard
- [x] Physical device using a non-Gboard keyboard

## Types of changes

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by AmandaRiu** [comment]
# Writing flow tests for [ENTER] behavior

These tests should be performed both with and without a Gboard keyboard. It may be worth paying close attention to the known issues with enter key behavior to make sure there aren't any new regressions, and if possible, whether some other issues are fixed by these changes.

## With Gboard keyboard

✅ Multiline components - Test Cases

 <details>
 <summary>Testing Steps</summary>

Test the next steps on:
- [x] Quote block
- [x] Verse block
- [x] Preformatted block
- [x] Code block (DEV only)
- [x] Pullquote block

##### ✅ TC001

**Known Issues**
- On Android, tapping Enter to split a quote or pullquote citation both splits the block and adds a newline to the citation ([#2498](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2498))
- On Android, pressing enter multiple times in the content of a Pullquote or Quote block causes unintended behavior ([#27690](https://github.com/WordPress/gutenberg/issues/27690))
- On Android, there is sometimes a loss of lines when entering multiple lines. ([#29861](https://github.com/WordPress/gutenberg/issues/29861))
- There is sometimes inconsistent HTML surrounding newlines in the Pullquote and Quote Block.  ([#1396](https://github.com/WordPress/gutenberg/issues/1396))

**New line on multiline components**

- Start typing and press enter to create a new line.
- The new line should be created, without splitting the block.
- On citations, pressing Enter at the end of a citation splits the block, but pressing enter in the middle of a citation creates a new line.
- Check on HTML mode that the resulting HTML code is correct:
  - *Quote*: `<p>` tags per paragraph.
  - *Quote citation*: `<br>` ending lines.
  - *Verse*: `<br>` ending lines.
  - *Preformatted*: `<br>` ending lines.
  - *Code*: (Invisible `\n`) new line character.
  - *Pullquote*: `<p>` tags per non-wrapping "line"
  - *Pullquote citation*: `<br>` ending lines.

 </details>

 ---

✅  Rich Text Formatting - Test Cases

 <details>
 <summary>Testing Steps</summary>

#### **Precondition**

Have a rich-text based component with content (Paragraph, Heading, Quote, Media Caption, etc...)


##### ✅ TC003

**Alignment Split**

- Write some lines on a paragraph block.
- Set an alignment on the block.
- Move the caret by the half of the text and press enter to split it.
- Check that the new block has the same alignment.
- Check that the text alignment corresponds to the setting.


 </details>

 ---

❌  Splitting and merging - Test Cases

 <details>
 <summary>Testing Steps</summary>


**Known Issues**
- ✅ **[Android]** Cannot split blocks immediately after prepending to a "swiped" word https://github.com/wordpress-mobile/gutenberg-mobile/issues/2372 - **this issue seems to be resolved**
-  **[Android]** Splitting block before "swiped" word loses "swiped" word https://github.com/wordpress-mobile/gutenberg-mobile/issues/2373

#### **Precondition**

Start from an empty post.

*Initial steps*

- Repeat the next steps on a Paragraph and Heading block
- Write (or paste) a paragraph with a few lines
- Set the caret by the half of the paragraph
- Press Enter (Intro) to split the block into two.


##### ✅ TC001

**Known Issues**

- **[Android]** Deleting "swiped" or selected words can prevent merging two blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375

**Merge after writing**
- Follow the initial steps.
- Write some more text where the caret is placed (beginning of new block).
- Delete all new text using the delete button until the blocks are merged again.
- Check that the blocks were merged.


##### ❌ TC002

**Known Issues**
- **[Android]** Deleting "swiped" or selected words prevents merging blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375

**Merge after selection**
- Follow the initial steps.
- Write some more text where the caret is placed (beginning of new block)
- Select all the newly written text,
- Press delete to remove all selected text.
- Press delete once again to merge the blocks. - **Unable to merge after selection was deleted**
- Check that the blocks were merged.


##### ✅  TC003

**Known Issues**
- **[Android]** Deleting "swiped" or selected words prevents merging blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375

**Merge after deleting text**
- Follow the initial steps.
- Move the caret a few words.
- Delete all those words until the blocks merge.
- Check that the blocks were merged.


##### ✅ TC004

**Known Issues**
- **[Android]** Deleting "swiped" or selected words prevents merging blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375

**Merge after deleting all**
- Follow the initial steps.
- Select and remove all content of the new block
- Press delete to remove the empty block.
- Check that the previous block was selected.


##### ✅ TC005

**Merge multiple blocks**
- Follow the initial steps.
- Press enter two times to create an empty block in between
- Press back two times to merge everything again.
- Check that the blocks were merged.


##### ✅ TC006

**Splitting/merge list block** 
- Write some items on a list block
- Put the caret by the half of them (at the end of the item) and press Enter to create a new item.
- Press Enter again to split the block into two, with an empty paragraph block in the middle. The paragraph block at the middle should be focused.
- Write something on that newly created paragraph block.
- Move the cursor to the beginning of the paragraph block
- Press delete to merge the paragraph block with the upper list block.
- Check that the new word should be a new list item.
- Move the caret to the beginning of the second list block.
- Press delete once to merge it back into the first list block.
- Check that everything was merged and there’s just one list block.

 </details>

## With non-Gboard keyboard

✅ Multiline components - Test Cases

 <details>
 <summary>Testing Steps</summary>

Test the next steps on:
- [x] Quote block
- [x] Verse block
- [x] Preformatted block
- [x] Code block (DEV only)
- [x] Pullquote block

##### TC001

**Known Issues**
- ☑️ On Android, tapping Enter to split a quote or pullquote citation both splits the block and adds a newline to the citation ([#2498](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2498)) - **Not an issue using the swift keyboard**
- ☑️ On Android, pressing enter multiple times in the content of a Pullquote or Quote block causes unintended behavior ([#27690](https://github.com/WordPress/gutenberg/issues/27690)) - **not an issue with the swift keyboard**
- ☑️ On Android, there is sometimes a loss of lines when entering multiple lines. ([#29861](https://github.com/WordPress/gutenberg/issues/29861)) - **not an issue with the swift keyboard**
- ☑️ There is sometimes inconsistent HTML surrounding newlines in the Pullquote and Quote Block.  ([#1396](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1396)) - **not an issue with the swift keyboard**

**New line on multiline components**

- Start typing and press enter to create a new line.
- The new line should be created, without splitting the block.
- On citations, pressing Enter at the end of a citation splits the block, but pressing enter in the middle of a citation creates a new line.
- Check on HTML mode that the resulting HTML code is correct:
  - *Quote*: `<p>` tags per paragraph.
  - *Quote citation*: `<br>` ending lines.
  - *Verse*: `<br>` ending lines.
  - *Preformatted*: `<br>` ending lines.
  - *Code*: (Invisible `\n`) new line character.
  - *Pullquote*: `<p>` tags per non-wrapping "line"
  - *Pullquote citation*: `<br>` ending lines.

 </details>

 ---

✅  Rich Text Formatting - Test Cases

 <details>
 <summary>Testing Steps</summary>

#### **Precondition**

Have a rich-text based component with content (Paragraph, Heading, Quote, Media Caption, etc...)


##### ✅ TC003

**Alignment Split**

- Write some lines on a paragraph block.
- Set an alignment on the block.
- Move the caret by the half of the text and press enter to split it.
- Check that the new block has the same alignment.
- Check that the text alignment corresponds to the setting.


 </details>

 ---

❌  Splitting and merging - Test Cases

 <details>
 <summary>Testing Steps</summary>


**Known Issues**
-  ☑️ **[Android]** Cannot split blocks immediately after prepending to a "swiped" word https://github.com/wordpress-mobile/gutenberg-mobile/issues/2372 - **not an issue with swift keyboard**
- ☑️  **[Android]** Splitting block before "swiped" word loses "swiped" word https://github.com/wordpress-mobile/gutenberg-mobile/issues/2373 - **not an issue on the swift keyboard**

#### **Precondition**

Start from an empty post.

*Initial steps*

- Repeat the next steps on a Paragraph and Heading block
- Write (or paste) a paragraph with a few lines
- Set the caret by the half of the paragraph
- Press Enter (Intro) to split the block into two.


##### ✅ TC001

**Known Issues**

- **[Android]** Deleting "swiped" or selected words can prevent merging two blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375

**Merge after writing**
- Follow the initial steps.
- Write some more text where the caret is placed (beginning of new block).
- Delete all new text using the delete button until the blocks are merged again.
- Check that the blocks were merged.


##### ❌ TC002

**Known Issues**
- ☑️ **[Android]** Deleting "swiped" or selected words prevents merging blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375 - **not an issue using the swift keyboard**

**Merge after selection**
- Follow the initial steps.
- Write some more text where the caret is placed (beginning of new block)
- Select all the newly written text,
- Press delete to remove all selected text.
- Press delete once again to merge the blocks.
- Check that the blocks were merged.


##### ✅ TC003

**Known Issues**
- ☑️ **[Android]** Deleting "swiped" or selected words prevents merging blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375 - **not an issue using the swift keyboard**

**Merge after deleting text**
- Follow the initial steps.
- Move the caret a few words.
- Delete all those words until the blocks merge.
- Check that the blocks were merged.


##### ✅ TC004

**Known Issues**
- **[Android]** Deleting "swiped" or selected words prevents merging blocks https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375

**Merge after deleting all**
- Follow the initial steps.
- Select and remove all content of the new block
- Press delete to remove the empty block.
- Check that the previous block was selected.


##### ✅ TC005

**Merge multiple blocks**
- Follow the initial steps.
- Press enter two times to create an empty block in between
- Press back two times to merge everything again.
- Check that the blocks were merged.


##### ✅ TC006

**Splitting/merge list block**

- Write some items on a list block
- Put the caret by the half of them (at the end of the item) and press Enter to create a new item.
- Press Enter again to split the block into two, with an empty paragraph block in the middle. The paragraph block at the middle should be focused.
- Write something on that newly created paragraph block.
- Move the cursor to the beginning of the paragraph block
- Press delete to merge the paragraph block with the upper list block.
- Check that the new word should be a new list item.
- Move the caret to the beginning of the second list block.
- Press delete once to merge it back into the first list block.
- Check that everything was merged and there’s just one list block.

 </details>

**Comment by AmandaRiu** [comment]
This is ready for review, but hold off on merging. I'm waiting to hear if we are okay with fixing this bug at the cost of test **TC002** in the **Splitting and merging** suite of tests failing. (cc: @mchowning @hypest )

**Comment by mkevins** [comment]
Hi @AmandaRiu :wave: :smile: 

I noticed that these changes appear to come from the first approach attempted on the original draft PR, but I may have left the description of that PR in the state from the second attempted approach there, even though those changes had been reverted. I think I should have changed the description back when I pushed the reverts, which would be:

> ## Description
> 
> The approach used in this PR adds logic to detect when we've encountered the unusual behavior exhibited by Gboard which leads to a failure to detect onEnter keypresses. When such behavior is detected, it calculates new offsets to be used in string comparison and processing to conform to the original intention of this TextWatcher.
> 
> Note: This is probably not enough, since this only addresses when the caret is at the end of the word, but Gboard may exhibit the unusual behavior also when the caret is in the middle or start of the word. Some consideration should be made for those other conditions as well.

In [my tests there](https://github.com/WordPress/gutenberg/pull/30368#issuecomment-816434758), this more "targeted" approach had not introduced any new regressions, and resolved the single issue of pressing enter at the end of a suggested ("underlined") word.

I also observed at that time that a few of the known issues had been resolved on `trunk`, so it is quite possible that the writing flow tests may exhibit different behavior now since the changeset is now rebased. I just wanted to clarify, since while in it's draft state that older PR had two _different_ (though similar) approaches attempted, and it looks like the changeset here corresponds with the original approach used there.

**Comment by hypest** [comment]
> The test that failed was the one where you first split a paragraph block, then add text to the beginning of the second block and then select that new text and delete it. Tapping delete again will not merge the two blocks back into a single block.

👋 @AmandaRiu , thanks for the thorough testing! Let me ask, will the user eventually be able to merge the blocks? Like, doing another backspace or maybe moving the caret somewhere and then the block would allow merging? If yes, then it feel that's a lesser issue than the newline-on-enter and I think it'd be useful to go ahead and land the fix for the newline issue anyway. We can then work on the merge issue separately.

**Comment by AmandaRiu** [comment]
> Hi @AmandaRiu 👋 😄
> 
> I noticed that these changes appear to come from the first approach attempted on the original draft PR, but I may have left the description of that PR in the state from the second attempted approach there, even though those changes had been reverted. I think I should have changed the description back when I pushed the reverts, which would be:
> 
> > ## Description
> > The approach used in this PR adds logic to detect when we've encountered the unusual behavior exhibited by Gboard which leads to a failure to detect onEnter keypresses. When such behavior is detected, it calculates new offsets to be used in string comparison and processing to conform to the original intention of this TextWatcher.
> > Note: This is probably not enough, since this only addresses when the caret is at the end of the word, but Gboard may exhibit the unusual behavior also when the caret is in the middle or start of the word. Some consideration should be made for those other conditions as well.
> 
> In [my tests there](https://github.com/WordPress/gutenberg/pull/30368#issuecomment-816434758), this more "targeted" approach had not introduced any new regressions, and resolved the single issue of pressing enter at the end of a suggested ("underlined") word.
> 
> I also observed at that time that a few of the known issues had been resolved on `trunk`, so it is quite possible that the writing flow tests may exhibit different behavior now since the changeset is now rebased. I just wanted to clarify, since while in it's draft state that older PR had two _different_ (though similar) approaches attempted, and it looks like the changeset here corresponds with the original approach used there.

Thanks @mkevins, I've updated the description in the PR. 👍

**Comment by AmandaRiu** [comment]
> > The test that failed was the one where you first split a paragraph block, then add text to the beginning of the second block and then select that new text and delete it. Tapping delete again will not merge the two blocks back into a single block.
> 
> 👋 @AmandaRiu , thanks for the thorough testing! Let me ask, will the user eventually be able to merge the blocks? Like, doing another backspace or maybe moving the caret somewhere and then the block would allow merging? If yes, then it feel that's a lesser issue than the newline-on-enter and I think it'd be useful to go ahead and land the fix for the newline issue anyway. We can then work on the merge issue separately.

that's a good question @hypest . I'll test it out and post the results here.

**Comment by AmandaRiu** [comment]
@hypest 

> Let me ask, will the user eventually be able to merge the blocks? Like, doing another backspace or maybe moving the caret somewhere and then the block would allow merging?

Hitting backspace again doesn't fix it, but if I click off the block and then click back into it, I'm able to backspace and merge the blocks back together.

**Comment by mkevins** [comment]
@AmandaRiu  :wave: :smile:  thank you for testing these flows so thoroughly!

Regarding TC002, I was wondering if you are observing that as a _new_ regression (i.e. that it was introduced by these changes). When I tested this, it seemed to be an issue on `trunk` as well, and just now tested the same flow on `17.5-rc-1` and observed the same bug with merging being prevented:

<video src="https://user-images.githubusercontent.com/8507675/121121742-ae580200-c80f-11eb-9cf2-1674a4dc2b83.mp4" width="360"></video>

Are you able to reproduce this on a current release as well?

**Comment by AmandaRiu** [comment]
@mkevins 

> Regarding TC002, I was wondering if you are observing that as a new regression (i.e. that it was introduced by these changes). When I tested this, it seemed to be an issue on trunk as well, and just now tested the same flow on 17.5-rc-1 and observed the same bug with merging being prevented:

The video you provided in you comment shows you were testing with "swiping" words, which is a known issue logged in this ticket https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375. The issue I had uncovered was just tapping out the word one letter at a time which wasn't logged as a known issue which is why I flagged it as a regression. I just tried the same scenario on the current production version of WPAndroid, and it IS happening there so this isn't a regression. I didn't think to test that before. With that in mind I think this PR is good to go 👍 

Unfortunately, since I'm the PR author, I'm unable to approve this PR so I will need someone else to do it so it may be merged.

**Comment by AmandaRiu** [comment]
I'll update the current ticket (https://github.com/wordpress-mobile/gutenberg-mobile/issues/2375) with notes that it happens with swiping and tapping out words.

**Comment by mkevins** [comment]
> The issue I had uncovered was just tapping out the word one letter at a time

Ah, good distinction! And thanks for confirming the behavior also exists for tapping!

**Review by mkevins — APPROVED** [review]
Approving via @AmandaRiu 's testing and review :smile:

**Pull request merged**
